File: New folder/cpp/mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int arr[],int l,int mid,int r){
    int n1=mid-l+1;
    int n2=r-mid;
    int arr1[n1],arr2[n2];

    for(int i=0;i<n1;i++){
        arr1[i]=arr[l+i];
    }
    for(int i=0;i<n2;i++){
        arr2[i]=arr[mid+i+1];
    }

    int i=0,j=0,k=l;
    while(i<n1 && j<n2){
        if(arr1[i]<arr2[j]){
            arr[k++]=arr1[i++];
        }
        else{
            arr[k++]=arr2[j++];
        }
    }
    while(i<n1){
        arr[k++]=arr1[i++];
    }
    while(j<n2){
        arr[k++]=arr2[j++];
    }
}
void mergeSort(int arr[],int l,int r){
    if(l<r){
        int mid=(l+r)/2;
        mergeSort(arr,l,mid); 
        mergeSort(arr,mid+1,r); 
        merge(arr,l,mid,r); 
    }
}
int main()
{
    cout << "Enter the number of elements -> ";
    int n; 
    cin>>n;
    cout << "Enter the elements -> ";
    int * arr=new int[n];
    for(int i =0;i<n;i++) cin>> arr[i];
    cout << "Before sorting -> ";
    for(int i=0;i<n;i++)
        cout << arr[i] << " ";
    cout << endl;
    mergeSort(arr,0,n-1);
    cout << "After sorting -> ";
    for(int i=0;i<n;i++) cout << arr[i] << " ";
    cout << endl;

    delete []arr;
    return 0;
}

File: python/mergeSort.py
def merge(arr,mid,l,r):
    n1=mid-l+1
    n2=r-mid
    a,b=[0]*n1,[0]*n2
    for i in range(n1):
        a[i]=arr[l+i]
    
    for i in range(n2):
        b[i]=arr[mid+i+1]

    i,j,k=0,0,l
    while i<n1 and j<n2:
        if a[i]>b[j]:
            arr[k]=b[j]
            j+=1
            k+=1
        else:
            arr[k]=a[i]
            i+=1
            k+=1
    while i<n1:
        arr[k]=a[i]
        k+=1
        i+=1
    while j<n2:
        arr[k]=b[j]
        k+=1
        j+=1

def mergeSort(arr,l,r):
    if l<r:
        mid=(l+r)//2
        mergeSort(arr,l,mid)
        mergeSort(arr,mid+1,r)
        merge(arr,mid,l,r)
if __name__=="__main__":
    n = int(input("Enter the size of the array -> "))
    arr=list(map(int,input("Enter the elements -> ").strip().split()))
    print("Before sorting ->",arr)
    mergeSort(arr,0,n-1)
    print("After sorting ->",arr)